Add HHAST migration to convert PHP stdlib calls to HSL

ssandler opened the pull request:
Posting this to start a discussion about how to make this ready for wider use. We're using this to migrate functions from PHP standard library calls to HSL. It's got some rough edges, but has been useful enough to run against much of the codebase without major issues.

Things on my mind to improve:

1. Clean up `getUseDeclaration` to automatically group multiple HSL namespaces. Also is generating the code as a string ok here? I originally did it with the AST generating each node at a time... but just generating one line of code required instantiating 49 classes, so readability suffers.
2. Detect the right argument order for `implode()` conversions, since that's ambiguous
3. Add more functions to this, if any are missing
4. Automatically detect whether to use `Str\replace_every()` instead of `Str\replace()` depending on whether the the args are KeyedContainers vs. strings. Currently I've been fixing these cases by hand after running the migrator.
5. Add unit tests
6. Make the bits that detect null vs. false comparisons smarter. Possibly integrate with `hh_client` and look for type errors that can be fixed by changing `false` to `null`. Right now it works nicely if the result of the function call like `Str\search` is compared with `false` immediately, but if it's saved off and used elsewhere it has to be fixed by hand. Luckily, the new hh_client errors about always-false expressions help a ton with that.
7. Detect negative length arguments to `Str\slice` if possible, since those throw exceptions. That's been a sticking point for us which we've had to manually keep an eye out for.
8. Whitespace formatting on this file. I ran `hackfmt` with no other args. Is that the norm? The formatting on `PHP_HSL_REPLACEMENTS` is inconsistent from one element to the next. Perhaps I should bump up the line width so it puts all the configs on one line?
9. Insert your ideas here

ssandler commented:
Just pushed a whole lot of updates, addressing all the issues I posted above, and a bunch more:

- Added unit tests. I put a bunch of echo commands in these example files, and verified that when actually executing the code the "expected" and "in" file produce the same results (other than `array` vs. `vec`).
- Automatically group namespace use declarations
- Use hh_client to find the types of nodes, and use that to fix the argument order for `implode()` which is ambiguous, as well as `str_replace` with array vs. string inputs
- Rewrite negative length arguments to `Str\slice()` so that they don't throw runtime exceptions. This was harder to do for `Str\splice()` and we don't use the latter much, so I didn't handle negative lengths for splice, just skip those invocations if found.
- Add a whole lot of new functions, covering everything I could find in the HSL reference that has a straightforward equivalent
- `min()` vs. `minva()` based on argument count, and same for max/maxva

I think this is about as much functionality as I want on this initial PR. More ideas for future improvements:
- batch up the requests to `hh_client`, since it can get slow if the few functions we use it for are called a lot in the file.
- handle other built-ins that are harder to convert over, like strncmp
- handle `isset()`, `empty()` in some intelligent way
- tie this code up to an auto-fixing linter
- convert `in_array()`, `array_key_exists()` to use `C\contains()`, `C\contains_key()`

fredemmott commented on src/find_type_for_node.php:
```diff
@@ -0,0 +1,78 @@
+namespace Facebook\HHAST\__Private;
+;
+
+
```
Delete one of these blank lines. https://github.com/hhvm/hhast/issues/109

fredemmott commented on src/find_type_for_node.php:
```diff
@@ -0,0 +1,78 @@
+ */
+
+namespace Facebook\HHAST\__Private;
+;
```
Delete. #110

fredemmott commented on src/find_type_for_node.php:
```diff
@@ -0,0 +1,78 @@
+  EditableNode $node,
+  string $path,
+): ?string {
+
```
delete (#111)

fredemmott commented on src/find_type_for_node.php:
```diff
@@ -0,0 +1,78 @@
+    \escapeshellarg($path),
+  ];
+  $output = \tempnam(\sys_get_temp_dir(), 'hhast-temp');
+  \exec(Str\join($command, ' ').' >'.\escapeshellarg($output));
```
make this function async, and use/extend `__Private\execute_async` if possible

ssandler replied:
Resolved! I had to make some changes due to the escaping. It was not possible to do input redirection due to the escaping built into `execute_async`, and I was having trouble getting `hh_client` to not complain when passed the path to the file (`hh_client --type-at-pos /tmp/file:line:offset` would complain about hh_server being stopped).

Turns out it wanted a path to a directory as a second arg, adding that allowed me to pass the file path and the containing directory, skipping input redirection. i.e. i.e. `hh_client --type-at-pos /tmp/file:line:offset /tmp`

fredemmott commented on src/find_type_for_node.php:
```diff
@@ -0,0 +1,78 @@
+  );
+
+  if (Shapes::keyExists($data, 'full_type')) {
+    if ($data['full_type']['kind'] === 'primitive') {
```
not a requirement, but altenratively:

```
if (($data['full_type']['kind'] ?? null) === 'primitive')
```

ssandler replied:
Thanks, took this suggestion

fredemmott commented on src/find_type_for_node.php:
```diff
@@ -0,0 +1,78 @@
+    }
+    return $data['full_type']['kind'];
+  }
+  ;
```
delete

fredemmott commented on src/find_type_for_node.php:
```diff
@@ -0,0 +1,78 @@
+    $untyped_data,
+  );
+
+  if (Shapes::keyExists($data, 'full_type')) {
```
We like to reduce nesting - if you don't take the `??` approach:

```Hack
if (!Shapes::keyExists($data, 'full_type')) {
  return null;
}
```

ssandler replied:
makes sense. Used the `??` approach

fredemmott commented on tests/fixtures/migrations/hsl.php.expect:
```diff
@@ -0,0 +1,127 @@
+    Str\replace_every('a test example with arrays', Dict\associate(['a', 'b'], ['x', 'y']));
+  $count = 0;
+  // this case is not handled in HSL
+  echo "\nstr_replace: ".str_replace('foo', 'bar', 'foobarbaz', &$count);
```
Is this because of the `&count` ? If so, can you put that in the comment

ssandler replied:
Yes, commented.

fredemmott commented on tests/fixtures/migrations/hsl.php.expect:
```diff
@@ -0,0 +1,127 @@
+  // this case is not handled in HSL
+  echo "\nstr_replace: ".str_replace('foo', 'bar', 'foobarbaz', &$count);
+  echo "\nstr_ireplace: ".Str\replace_ci('foobarbaz', 'foo', 'bar');
+  // this case is also not handled
```
also extend this comment please, and the others below

ssandler replied:
Done

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,646 @@
+
+use function Facebook\HHAST\__Private\find_type_for_node;
+
+enum HSL_NAMESPACE: string as string {
```
is the `as string` required?

ssandler replied:
Nope! I removed it and put `coerce()` in a few other places.

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,646 @@
+  Vec = 'Vec';
+}
+
+type hsl_replacement_config = shape(
```
`UpperCamelCase` like class names: #112

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,646 @@
+  Keyset = 'Keyset';
+  Math = 'Math';
+  Str = 'Str';
+  Vec = 'Vec';
```
uppercase names (#113)

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,646 @@
+    $nodes = $root->getDescendantsOfType(FunctionCallExpression::class);
+
+    // keep track of any HSL namespaces we may have to add to the top of the file
+    $found_namespaces = dict[];
```
Why a dict rather than a `keyset` ?

ssandler replied:
Good point, fixed! I remembered keysets could only contain strings, and didn't think about the fact that the enum was ok to use in a keyset. This simplified a lot of logic.

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,646 @@
+
+    // add "use namespace" declarations at the top if they aren't already present
+    $declarations =
+      vec($root->getDescendantsOfType(INamespaceUseDeclaration::class));
```
This is ok for the common case, but:

```Hack
namespace Foo {
}

namespace Bar {
  use namespace HH\Lib\Vec;
}
```

... is also valid

ssandler replied:
Should be addressed in https://github.com/hhvm/hhast/pull/108/commits/ba302ff77b6cea2ba4aae34fe7710048400071d9

I cleaned up the logic, using `insertBefore` to skip over namespace declarations, and also handled both kinds of namespace declarations (with and without body).

I didn't do anything special to handle two block namespace declarations in the same file, though. It will always insert it inside the first one.

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,646 @@
+    $declarations =
+      vec($root->getDescendantsOfType(INamespaceUseDeclaration::class));
+    list($hsl_declarations, $suffixes) =
+      $this->findUseDeclarations($declarations);
```
Might want to look at ./__Private/Resolution/get_uses_directly_in_scope.php
./__Private/Resolution/get_current_uses.php
./__Private/Resolution/get_current_namespace.php

ssandler replied:
Any specific changes you're looking for here? I did crib off of these functions a bit, but I don't think I can use any of them directly?

fredemmott replied:
Nope, just wanted to make sure you were aware of them, and that this needed doing here

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,646 @@
+    list($hsl_declarations, $suffixes) =
+      $this->findUseDeclarations($declarations);
+
+    $count_before = C\count(Vec\unique($suffixes));
```
efficiency: replace `Vec\unique($foo)` with `keyset($foo)` here: `Vec\unique($x)` is basically `vec(keyset($x))`

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,646 @@
+
+    // add new suffixes to the current list of suffixes
+    $suffixes = Vec\concat($suffixes, Vec\keys($found_namespaces))
+      |> Vec\unique($$);
```
would a keyset generally be more appropriate throughout?

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,646 @@
+      |> Vec\unique($$);
+
+    // added any new suffixes?
+    if (C\count($suffixes) !== $count_before) {
```
Reduce indentation: if ===, return early

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,646 @@
+      $node->getOperator() instanceof MinusToken
+    ) {
+      $val = $this->resolveIntegerArgument($node->getOperand());
+      return $val !== null ? -1 * $val : null;
```
Please add brackets - this is fine, but `? :`'s weirdness in PHP makes me nervous :p

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,646 @@
+      foreach ($arguments as $i => $argument) {
+        invariant($argument instanceof ListItem, 'expected ListItem');
+        $new_argument_list[] =
+          $argument->replace($argument->getItem(), $new_items[(int)$i]);
```
Is there a test covering this, both with and without trailing commas?

e.g, we can't turn `foo(1, 2, 3)` into `bar(3 1, 2,)` ?

ssandler replied:
Added a test for this (the commas stay exactly where they are)

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,646 @@
+      }
+
+      if ($decl instanceof NamespaceGroupUseDeclaration) {
+
```
delete

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,646 @@
+          if ($name instanceof NameToken) {
+            $suffixes[] = $name->getText();
+          }
+        }
```
This is fine - but alternatively (up to you):

```Hack
$suffixes = $clauses
  |> Vec\map($$, $c ==> { $n = $c->getName(); return $n instanceof NameToken ? $n->getText(): null; })
  |> Vec\filter_nulls($$)
  |> Vec\concat($suffixes, $$);
```

This is also another case where it feels like `keyset[]` is the right approach

ssandler replied:
I used this suggestion, thanks!

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,646 @@
+      // single namespace use declaration
+      $ns = C\firstx($suffixes);
+    }
+    return $this->nodeFromCode(
```
https://github.com/hhvm/hhast/issues/6 would be nice, but could be future work to clean this up, rather than a blocker.

ssandler replied:
That would be awesome (and as I just learned, it'd be good to be able to do that with insertBefore and insertAfter too). I'd feel much better about using a code string if it were part of the library :-).

Let's address separately though since this PR is getting large.

fredemmott commented on src/__Private/MigrationCLI.php:
```diff
@@ -21,6 +21,7 @@
   ImplicitShapeSubtypesMigration,
   OptionalShapeFieldsMigration,
   NamespaceFallbackMigration,
+  HSLMigration
```
Trailing comma

ssandler commented:
I believe all comments are addressed. I also added tests for the in and expect files having the same output.

## Running the migrator on hhast itself

As suggested outside of this review, I also ran the migration on hhast itself with `bin/hhast-migrate --hsl src`, which identified the issue with leading slashes that I handled in  https://github.com/hhvm/hhast/pull/108/commits/e3e16c6be696aead033c4825af4f839d0528102f. I verified tests pass on after running that migration. I can commit the results of that to this PR if you'd like.

```
SFO-M-SSANDLER01:hhast ssandler$ bin/hhast-migrate --hsl src
SFO-M-SSANDLER01:hhast ssandler$ git status
On branch ssandler_hsl_migration
Your branch is ahead of 'origin/master' by 23 commits.
  (use "git push" to publish your local commits)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   src/Linters/DontAwaitInALoopLinter.php
	modified:   src/__Private/SubprocessException.php
	modified:   src/__Private/codegen/CodegenSyntax.php
	modified:   src/nodes/EditableNode.php
	modified:   src/nodes/EditableToken.php
	modified:   src/nodes/EditableTrivia.php
	modified:   src/resolve_type.php

no changes added to commit (use "git add" and/or "git commit -a")
SFO-M-SSANDLER01:hhast ssandler$ vendor/bin/phpunit 
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

.....................................................I.........  63 / 295 ( 21%)
............................................................... 126 / 295 ( 42%)
............................................................... 189 / 295 ( 64%)
............................................................... 252 / 295 ( 85%)
...........................................                     295 / 295 (100%)

Time: 1.49 minutes, Memory: 64.06MB

OK, but incomplete, skipped, or risky tests!
Tests: 295, Assertions: 375, Incomplete: 1.
```

fredemmott commented on tests/MigrationsTest.php:
```diff
@@ -158,4 +150,17 @@ classname<Migrations\BaseMigration> $migration,
+      \HH\Asio\join(HHAST\__Private\execute_async('hhvm', $base_path.'.in'));
+    $hsl_results = \HH\Asio\join(
+      HHAST\__Private\execute_async('hhvm', $base_path.'.expect'),
+    );
```
This should be a single HH\Asio\join, around a HH\Lib\Tuple\from_async

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,684 @@
+}
+
+type HslReplacementConfig = shape(
+  // \HH\Lib\* namespace the replacement function lives in
```
These should be `/** */` comments (supported by `hh-apidocs`)

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,684 @@
+    // build a possibly grouped namespace use declaration
+    $new_namespace_use_declaration = $this->buildUseDeclaration($suffixes);
+
+    // insert the nmew node: skip the <?hh sigil and namespace declaration if present,
```
nmew

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,684 @@
+      return $root->insertBefore($child, $new_namespace_use_declaration);
+    }
+
+    invariant(false, 'should not fail to insert new node');
```
invariant_violation

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,684 @@
+      }
+
+      return null;
+    } elseif (
```
- space between `else` and `if`
- given the if block above returns, this could just be an `if`

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,684 @@
+
+    $items = Vec\map(
+      $arguments,
+      ($argument) ==> {
```
parens uneeccessary

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,684 @@
+    // if neither arg is a string, hh_client should complain so we just leave it as is
+    $fn_name = $this->getFunctionName($node);
+    if ($fn_name === 'Str\\join') {
+      $type = \HH\Asio\join(find_type_for_node_async($root, $items[1], $path));
```
make this function async, await result

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,684 @@
+      if ($type === 'string') {
+        $argument_order = Vec\reverse($argument_order ?? vec[]);
+      }
+    } elseif ($fn_name === 'Str\\replace' || $fn_name === 'Str\\replace_ci') {
```
else if

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,684 @@
+      // str_replace and str_ireplace have two modes:
+      // string for search/replace args means replacing a single pattern
+      // arrays mean replacing a set of patterns, which we should rewrite as Str\replace_every
+      $type = \HH\Asio\join(find_type_for_node_async($root, $items[0], $path));
```
await

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,684 @@
+    $stack_count = C\count($stack);
+    $parent = $stack[$stack_count - 2];
+
+    if ($parent instanceof BinaryExpression) {
```
this also looks like a good candidate for negate the check, return early, unindent the rest

ssandler commented:
Just pushed fixes to all of those!

fredemmott commented on src/Migrations/HSLMigration.php:
```diff
@@ -0,0 +1,711 @@
+
+
+  <<__Override>>
+  public function migrateFile(string $path, EditableNode $root): EditableNode {
```
heh, I should change this to be an async function in the next version. THis is fine :)
